src/ui/components/Overlay/index.js

```diff
@@ -1,80 +1,110 @@
+/* @flow */
+import invariant from 'invariant';
 import makeClassName from 'classnames';
 import * as React from 'react';
-import PropTypes from 'prop-types';
+import { compose } from 'redux';
+
+import withUIState from 'core/withUIState';
 
 import './styles.scss';
 
-export default class Overlay extends React.Component {
-  static propTypes = {
-    children: PropTypes.node,
-    className: PropTypes.string,
-    onEscapeOverlay: PropTypes.func,
-    visibleOnLoad: PropTypes.bool,
-  };
+// TODO: look into this as I don't see this working on the frontend
+// even in other envs.
+export const onClickBackground = (
+  event: SyntheticEvent<HTMLButtonElement>,
+  // eslint-disable-next-line no-use-before-define
+  _this: OverlayBase,
+) => {
+  if (_this.props.onEscapeOverlay) {
+    _this.props.onEscapeOverlay();
+  }
+  _this.hide();
+};
+
+type Props = {|
+  className?: string,
+  children: React.Element<any>,
+  id: string,
+  onEscapeOverlay?: Function,
+  visibleOnLoad?: boolean,
+  _onClickBackground: Function,
+|};
 
+type UIStateType = {|
+  visible: boolean,
+|};
+
+const initialUIState: UIStateType = { visible: false };
+
+type InternalProps = {|
+  ...Props,
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+|};
+
+export class OverlayBase extends React.Component<InternalProps> {
   static defaultProps = {
     visibleOnLoad: false,
+    _onClickBackground: onClickBackground,
   };
 
-  constructor(props) {
-    super(props);
-    this.state = { visible: props.visibleOnLoad };
-  }
+  componentWillReceiveProps(nextProps: InternalProps) {
+    const { uiState: uiStateOld } = this.props;
+    const { visibleOnLoad: visibleOnLoadNew } = nextProps;
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.visibleOnLoad !== undefined) {
-      this.setState({ visible: nextProps.visibleOnLoad });
+    if (
+      (visibleOnLoadNew && !uiStateOld.visible) ||
+      uiStateOld.visible !== nextProps.uiState.visible
+    ) {
+      this.props.setUIState({ visible: visibleOnLoadNew });
     }
   }
 
-  onClickBackground = () => {
```

Comment:
It looks like maybe you moved this to a prop just for testing? I am not sure you need it for testing (I explained why in the comments).